File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if(root==NULL)
        return {};
        queue<TreeNode*> q;
        q.push(root);
        vector<vector<int>> ans;
        bool leftToRight = true;
        while(!q.empty())
        {
            int sizeOfQ = q.size();
            
            vector<int> curr_row;
           

            for(int i=0;i<sizeOfQ;i++)
            {
                TreeNode* front = q.front();
                q.pop();

                curr_row.push_back(front->val);
             

                if(front->left)
                    q.push(front->left);
                 if(front->right)
                    q.push(front->right);
            
            
            }
            
            if(leftToRight)
            {
                ans.push_back(curr_row);
                leftToRight = !leftToRight;
            }
            else{
                reverse(curr_row.begin(),curr_row.end());
                ans.push_back(curr_row);
                leftToRight = !leftToRight;
            }
        }
        return ans;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []

        result = []

        queue = deque([root])

        flag = 0

        while queue:
            size_of_queue = len(queue)
            print(size_of_queue)
            curr_level_val = []

            for _ in range(size_of_queue):
                node = queue.popleft()
                curr_level_val.append(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            if flag:
                curr_level_val.reverse()
            result.append(curr_level_val)
            curr_level_val = []
            flag = not flag

        return result
                
            
